File: python/grokking/two_pointers/subarrays_with_product_less_than_target.py
from collections import deque


def find_subarrays_with_product_less_than_target(arr: list[int], target: int) -> list[list[int]]:
    result = []
    product = 1.0
    left = 0
    
    for right in range(len(arr)):
        product *= arr[right]
        
        while product >= target and left < len(arr):
            product /= arr[left]
            left += 1
            
        temp_list = deque()
        
        for i in range(right, left - 1, -1):
            temp_list.appendleft(arr[i])
            result.append(list(temp_list))

    return result

File: golang/two_pointers/subarrays_with_product_less_than_target.go
package two_pointers

func findSubarrays(arr []int, target int) [][]int {
	var result [][]int
	product := 1.0
	left := 0

	for right := 0; right < len(arr); right++ {
		product *= float64(arr[right])

		for product >= float64(target) && left < len(arr) {
			product /= float64(arr[left])
			left++
		}

		var tempList []int
		for i := right; i > left-1; i-- {
			tempList = append([]int{arr[i]}, tempList...)
			result = append(result, tempList)
		}
	}

	return result
}
